### ISSUE TITLE
Is it possible to download the new SF 'ContentDocument' object type through Restforce?

### ISSUE THREAD

**Issue by Arepo** [restforce/restforce#377]
As far as I can see, SF no longer uses the 'Attachment' object type for new object types. I recently set one up to have what I thought was an attachment and eventually realised it was a 'ContentDocument' - but I'm unsure how to download the actual file via the API since neither the ContentDocument object nor the associated ContentVersion or ContentDocumentLink objects appear to have a 'body' field that's directly writable to a file. 

The best I can find is a URL path, and I haven't figured out how to get the authorisation for it to work yet, so I'm hoping there's some simpler way.

**Comment by antwonlee** [comment]
Maybe this will help?

https://salesforce.stackexchange.com/questions/212488/retrieve-contentdocument-file-rest-api?answertab=active#tab-top

**ContentVersion**
https://developer.salesforce.com/docs/atlas.en-us.api.meta/api/sforce_api_objects_contentversion.htm

**Comment by Arepo** [comment]
Is there any plan to integrate the functionality into Restforce? We can do it the old-fashioned way if we have to, but would prefer something that works out of the box with the code we already have

**Comment by Arepo** [comment]
Just been looking at this again after putting it on hold for a few months. I'm not sure if this is a Restforce or SF issue or my not understanding how to use one of them, but following those guides only produces a url. That is, fetching a specific Content Version (with client.find('ContentVersion', '<version_id>')  and calling '#VersionData' on it gets me a string of the following sort: '/services/data/v38.0/sobjects/ContentVersion/0681o000003lVpcAAE/VersionData'

But within Restforce I'm not sure how to access the url, or whether I even should - it seems unlikely that I would have to send a query just to get another URL to query? So is this the expected behaviour? And if so, how do I interact with the URL?

**Comment by Arepo** [comment]
Relatedly, I'm also looking for a way to fetch all ContentDocuments attached to a specific record. I found [this thread](https://success.salesforce.com/answers?id=9063A000000l1gCQAQ) which mentions finding a LinkedEntity, but I can't see any such field via the Restforce interface - not sure if Restforce isn't relaying it, or it just isn't there?

**Comment by azharbeebeejaun** [comment]
@Arepo Did you figure out any way to access the URL? I'm stuck with the same problem. The ContentVersion field is a URL of which I'm unsure how to access. The same ContentVersion field is in base64 when queried from workbench...

**Comment by Arepo** [comment]
@azharbeebeejaun No, I had to shelve this one for the time being. Would love to see your solution!

**Comment by azharbeebeejaun** [comment]
Below's how it basically goes. I have a ContentDocument ID (0694I000008HV8MQAW).
```
client = Restforce.new(...)
content_version = client.query(
  "SELECT PathOnClient, VersionData FROM ContentVersion WHERE ContentDocumentId = '0694I000008HV8MQAW' AND IsLatest = true"
).first
name = content_version.PathOnClient
body = client.get(content_version.VersionData).body
# write body to a file
```

As for your related question about fetching all ContentDocuments attached to a specific record, you may use something like the SOQL below which will select all ContentDocuments on Case objects.
```
SELECT ContentDocumentId 
FROM ContentDocumentLink 
WHERE LinkedEntityId IN (SELECT Id FROM Case) AND LinkedEntity.Type = 'Case'
```

**Comment by Arepo** [comment]
Thanks! We'll give it a try

**Comment by StephRicardo** [comment]
Thanks for this info. It was really helpful for me. I am trying to create a ContentDocument with data instead of an Attachment. I was able to use the following: 

```bash
client.upsert!(
'ContentVersion', 
'Title', 
Title: 'test-attachment', 
PathOnClient: 'test-attachment.txt', 
FirstPublishLocationId: record_id, 
VersionData: body
)
```

By creating the ContentVersion and providing the information below, it creates a ContentDocument if one doesn't exist, and a ContentDocumentLink to the record id provided. This returns the id of the newly created ContentVersion that you can use to find the other related records.

`body` is base64 encoded string. `record_id` is the id of the object you want to link it to on Salesforce.

This source was also very helpful: https://salesforce.stackexchange.com/questions/262711/create-and-link-document-to-multiple-records-using-rest-api
